README.md
# mondrian-generator

Generate mondrian-like images.

## Animated Examples

![Example 1](./examples/ex1.gif)

## Examples

![Example 0](./examples/ex0.png)

![Example 1](./examples/ex1.png)

![Example 2](./examples/ex2.png)

![Example 3](./examples/ex3.png)

![Example 4](./examples/ex4.png)

![Example 5](./examples/ex5.png)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-01-19
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-01-19
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/mondrian-generator/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/mondrian-generator/compare/0.1.0...0.1.1


test/mondrian_generator/core_test.clj
(ns mondrian-generator.core-test
  (:require [clojure.test :refer :all]
            [mondrian-generator.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/mondrian_generator/core.clj
(ns mondrian-generator.core
  (:gen-class))

(defn random-color [options]  (rand-nth options))

(defn is-horizontal-division
  [section]
  (not (:is-horizontal section)))

(defn create-section [top, left, right, bottom, color, level is-horizontal]
  {:top top
   :left left
   :right right
   :bottom bottom
   :color color
   :level level
   :is-horizontal is-horizontal})


(defn create-horizontal-subsections [section division color-generator]
  (let [{:keys [top bottom left right level]} section
        new-level (inc level)
        line  (+ left (* division (- right left)))]
    {:subsections [(create-section
                    top left line bottom (color-generator) new-level true)
                   (create-section
                    top line right bottom (color-generator) new-level true)]
     :division [line top line bottom]}))

(defn create-vertical-subsections [section division color-generator]
  (let [{:keys [top bottom left right level]} section
        new-level (inc level)
        line  (+ top (* division (- bottom top)))]
    {:subsections [(create-section top left right line (color-generator) new-level false)
                   (create-section line left right bottom (color-generator) new-level false)]
     :division [left line right line]}))


(defn split-section [section division color-generator]
  (if (true? (is-horizontal-division section))
    (create-horizontal-subsections section division color-generator)
    (create-vertical-subsections section division color-generator)))

;; TODO: turn division, max-level and colors into a map
(defn- generate-mondrian-r
  [pending-sections sections divisions color-generator division-generator max-level]
  (if (> (count pending-sections) 0)
    (let [current-section (last pending-sections)
          division (division-generator current-section)]
      (if (or
           (= division 1)
           (= division 0)
           (>= (:level current-section) max-level))
        (recur
         (pop pending-sections)
         (conj sections current-section)
         divisions
         color-generator
         division-generator
         max-level)
        (let [{:keys [subsections division]} (split-section current-section division color-generator)]
          (recur
           (apply conj
                ; delete current section before adding the subsections
                  (pop pending-sections)
                ; new subsections
                  subsections)
           sections
           (conj divisions division)
           color-generator
           division-generator
           max-level))))
    {:sections sections :divisions divisions}))


(defn generate-mondrian [config division-generator color-generator max-level]
  (let [pending-sections
        (create-section 0 0 (:max-x config) (:max-y config) (color-generator) 0 false)]
    (generate-mondrian-r [pending-sections] [] [] color-generator division-generator max-level)))


(defn division-generator [section min-level]
  (let [{level :level} section]
    (if (< level min-level)
      (rand-nth [0.3 0.4 0.5 0.6 0.7])
      (rand-nth [0 0.3 0.4 0.5 0.6 0.7 1]))))

(defn- main []
  (generate-mondrian
   {:max-x 1000 :max-y 1000}
   (fn [section] (division-generator section 3))
   (fn [] (random-color ["red" "blue" "yellow" "white" "black"]))
   5))


src/mondrian_generator/draw.clj
(ns mondrian-generator.draw
  (:require [mondrian-generator.core :as core]
            [quil.core :as q]))


(def rgb {:red [255 0 0]
          :blue [66 135 245]
          :yellow [252 186 3]
          :black [0]
          :white [255]})

(defn draw-section [section]
  (let [{:keys [top left right bottom color]} section
        width (- right left)
        height (- bottom top)]
    (apply q/fill (rgb color))
    (q/rect left top width height)))

(defn draw []
  ; white background
  (q/no-loop)
  (q/stroke-weight 5)
  (q/background 255)
  (doseq [section
          (:sections
           (core/generate-mondrian
            {:max-x 1500 :max-y 1000}
            (fn [section] (core/division-generator section 2))
            ;; TODO: color generator will work in function of the contiguos section
            ;; prefer some color combinations over others
            (fn [] (core/random-color
                 ;; TODO: random colors given weight
                    [:red :blue :yellow :white :white :white :blue :yellow :white :white]))
            4))]
    (draw-section section))
  (q/save "generated/image.png")
  ;; comment to show window with output
  (q/exit))


(q/defsketch example
  :host "host"
  :title "Mondrianesque"
  ;; TODO: config map
  :size [1500 1000]
  :draw draw)


src/mondrian_generator/animate.clj
(ns mondrian-generator.animate
  (:require [quil.core :as q]
            [mondrian-generator.core :as core]
            [mondrian-generator.draw :as draw]
            [quil.middleware :as m]))

(defn setup []
  (q/background 255)
  (q/frame-rate 1)
  {:data (core/generate-mondrian
          {:max-x 1500 :max-y 1000}
          (fn [section] (core/division-generator section 2))
         ;; TODO: color generator will work in function of the contiguos section
         ;; prefer some color combinations over others
          (fn [] (core/random-color
                 ;; TODO: random colors given weight
                  [:red :blue :yellow :white :white :white :blue :yellow :white :white]))
          4)
   :shown []
   :index 0})

(defn draw [state]
  (let [{:keys [data index]} state
        {:keys [sections divisions]} data]
    (q/stroke-weight 5)
    (q/background 255)
    (loop [i 0]
      (when (< i (min (count divisions) index))
        (apply q/line (divisions i))
        (recur (inc i))))

    (when (<= (count divisions) index)
      (let [count (- index (count sections))]
        (loop [i 0]
          (when (< i count)
            (draw/draw-section (sections i))
            (recur (inc i))))))
    ;(q/save-frame (str "frames/frame-" index ".png"))
    (when (>= index (+ (count divisions) (count sections)))
      (q/exit))))


(defn update-state [state]
  (let [{:keys [data index]} state]
    {:data data
     :index (inc index)}))

(q/defsketch animate
  :host "host"
  :setup setup
  :title "Mondrianesque"
  :size [1500 1000]
  :update update-state
  :draw draw
  :middleware [m/fun-mode])

